model/riscv_insts_zicbom.sail

```diff
@@ -0,0 +1,110 @@
+// exist).
+
+function cbo_clean_flush_enabled(p : Privilege) -> bool = feature_enabled_for_priv(p, menvcfg.CBCFE()[0], senvcfg.CBCFE()[0])
+function cbo_inval_enabled(p : Privilege) -> bool = feature_enabled_for_priv(p, menvcfg.CBIE()[0], senvcfg.CBIE()[0])
+function cbo_inval_as_inval(p : Privilege) -> bool = feature_enabled_for_priv(p, menvcfg.CBIE()[1], senvcfg.CBIE()[1])
+
+/* ****************************************************************** */
+union clause ast = RISCV_ZICBOM : (cbop_zicbom, regidx)
+
+mapping encdec_cbop : cbop_zicbom <-> bits(12) = {
+  CBO_CLEAN <-> 0b000000000001,
+  CBO_FLUSH <-> 0b000000000010,
+  CBO_INVAL <-> 0b000000000000
+}
+
+mapping clause encdec = RISCV_ZICBOM(cbop, rs1)             if haveZicbom()
+  <-> encdec_cbop(cbop) @ rs1 @ 0b010 @ 0b00000 @ 0b0001111 if haveZicbom()
+
+mapping cbop_mnemonic : cbop_zicbom <-> string = {
+  CBO_CLEAN <-> "cbo.clean",
+  CBO_FLUSH <-> "cbo.flush",
+  CBO_INVAL <-> "cbo.inval"
+}
+
+mapping clause assembly = RISCV_ZICBOM(cbop, rs1)
+  <-> cbop_mnemonic(cbop) ^ spc() ^ "(" ^ opt_spc() ^ reg_name(rs1) ^ opt_spc() ^ ")"
+
+val process_clean_inval : (regidx, cbop_zicbom) -> Retired
+function process_clean_inval(rs1, cbop) = {
+  let rs1_val = X(rs1);
+  let cache_block_size_exp = plat_cache_block_size_exp();
+  let cache_block_size = 2 ^ cache_block_size_exp;
+
+  // Offset from the beginning of the cache block.
+  let offset = (rs1_val & ~(zero_extend(ones(cache_block_size_exp)))) - rs1_val;
+
+  // TODO: This is incorrect since CHERI only requires at least one byte
+  // to be in bounds here, whereas `ext_data_get_addr()` checks that all bytes
+  // are in bounds. We will need to add a new function, parameter or access type.
+  match ext_data_get_addr(rs1, offset, Read(Data), cache_block_size) {
+    Ext_DataAddr_Error(e) => { ext_handle_data_check_error(e); RETIRE_FAIL },
+    Ext_DataAddr_OK(vaddr) => {
+      let res: option(ExceptionType) = match translateAddr(vaddr, Read(Data)) {
+        TR_Address(paddr, _) => {
+          // "A cache-block management instruction is permitted to access the
+          // specified cache block whenever a load instruction or store instruction
+          // is permitted to access the corresponding physical addresses. If
+          // neither a load instruction nor store instruction is permitted to
+          // access the physical addresses, but an instruction fetch is permitted
+          // to access the physical addresses, whether a cache-block management
+          // instruction is permitted to access the cache block is UNSPECIFIED."
+          //
+          // In this implementation we currently don't allow access for fetches.
+          let exc_read = phys_access_check(Read(Data), cur_privilege, paddr, cache_block_size);
+          let exc_write = phys_access_check(Write(Data), cur_privilege, paddr, cache_block_size);
+          match (exc_read, exc_write) {
+            (Some(e0), Some(e1)) => Some(e1),
```

Comment:
Where is it specified that one should prefer the write exception?

Reply:
Also, if one fails but not the other, this treats that as a success? That seems odd.

Reply:
I guess that's from this:
> If access to the cache block is not permitted, a cache-block management instruction raises a store page fault or store guest-page fault exception if address translation does not permit any access or raises a store access fault exception otherwise.

Might be worth explaining that.

Reply:
Yep, comment added.